src/net/fe/overworldStage/ClientOverworldStage.java

```diff
@@ -653,24 +655,128 @@ public void setUnitInfoUnit(Unit u) {
 	
 	private void updateFog() {
 		if (fogOption != FogOption.NONE) {
-			Set<Node> nodes = Zone.all(grid);
-			for(Unit unit : getAllUnits())
-				if(FEMultiplayer.getLocalPlayer().getParty().isAlly(unit.getParty()) && unit.getHp() > 0)
-					for(int i = 0; i <= unit.getTheClass().sight; i++)
-						for(int j = 0; j <= unit.getTheClass().sight - i; j++) {
-							nodes.remove(new Node(unit.getOrigX() + i, unit.getOrigY() + j));
-							nodes.remove(new Node(unit.getOrigX() + i, unit.getOrigY() - j));
-							nodes.remove(new Node(unit.getOrigX() - i, unit.getOrigY() + j));
-							nodes.remove(new Node(unit.getOrigX() - i, unit.getOrigY() - j));
+			Set<Node> nodes = new HashSet<>();
+			if(FEMultiplayer.getLocalPlayer().isSpectator()) {
+				switch(session.getSpectatorFogOption()) {
+					case REVEAL_ALL:
+						//Nothing to do here
+						//nodes is already the empty set
+						break;
+					case MUTUALLY_VISIBLE_UNITS:
+						nodes = calcSpectatorFog();
+						break;
+					case SPECTATE_RED: {
+						nodes = Zone.all(grid);
+						Set<Player> players = new HashSet<Player>();
+						for(Player player : FEMultiplayer.getPlayers().values()) {
+							if(player.getParty().getColor().equals(Party.TEAM_RED)) {
+								players.add(player);
+								break;
+							}
 						}
+						for(Player player : players)
+							nodes.removeAll(Fog.getPartyPerception(player.getParty()));
+						break;
+					}
+					case SPECTATE_BLUE: {
+						nodes = Zone.all(grid);
+						Set<Player> players = new HashSet<Player>();
+						for(Player player : FEMultiplayer.getPlayers().values()) {
+							if(player.getParty().getColor().equals(Party.TEAM_BLUE)) {
+								players.add(player);
+								break;
+							}
+						}
+						for(Player player : players)
+							nodes.removeAll(Fog.getPartyPerception(player.getParty()));
+						break;
+					}
+				}
+			} else {
+				nodes = Zone.all(grid);
+				nodes.removeAll(Fog.getPartyPerception(FEMultiplayer.getLocalPlayer().getParty()));
+			}
 			fog.setNodes(nodes);
 		}
 	}
 	
+	/**
+	 * Çhanges the nodes parameter in order to match the mutual perception of each player's mutual perception (?)
+	 * Basically, spectators can only see the grid where all players know that all the others can see.
+	 * @param nodes The set of nodes to change
+	 */
+	private Set<Node> calcSpectatorFog() {
+		Set<Node> nodes = Zone.all(grid);
+		//This is not what God intended
```

Comment:
Given this comment, I feel like I must be missing something regarding `calcSpectatorFog`. Can you elaborate on what God does not intend?

Reply:
This is basically referring to the fact that the implementation of the function is sloppy. The name of the variables don't highlight the differences between their uses and they are pretty much used only once. There's also the fact that I lazily used a hash table for every set, which is probably not the fastest solution, but I really couldn't be bothered at the time. Also, _technically_ a R-tree could reduce the complexity from O(n^2) to O(n log n), I think? I have no idea how spacial indexing works and I'm fairly certain the standard Java library doesn't have that data structure. Anyway, this is a huge mess and I'm not proud of it, but it works and I've already spent enough time to get the thing to work. It's _somewhat_ future proof so hopefully no one will have to touch this code in the future.

Reply:
Hmm... reminds me of back in the day when my brother would tag code that wasn't how God intended either and eventually it came back to haunt him. When things slow down and you find yourself with nothing to do with this project, try to make it more pleasing to God.